---
title: "BAM_file_statistics"
author: "Hugo DENIS"
date: "2024-01-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Packages
```{r}
library(xlsx)
library(dplyr)
library(ggplot2)
library(stringr)
```

#### This scripts uses mapped BAM files to produce basic statistics on A.spathulata WGS data ####   
# This includes coverage, base quality, sequencing depth, mapping quality ...

#Load samples metadata (A.spathulata 2022 only)
```{r}
meta=read.csv('/home/hugo/PhD/GBR_Data/Coral colony metadata from the ECT1_RRAP_Aspathulata_2022.csv')
```

#Load BAM stats files  
```{r}
path='/home/hugo/PhD/Genomics/Raw_data_processing/BAM_statistics/'

#Trimmed statistics -> problem with disk 
df_trim=list.files("/data1/WGS_Aspat_GBR/Trimmed_files/","*sum.txt",full.names = T) %>% lapply(function(x)read.table(x,sep=":",col.names=c("Var","Value")) %>% mutate(sample.name=sub("_sum.*", "",basename(x)))) %>% bind_rows %>% mutate(Value=as.numeric(str_remove(str_replace(Value,",",".")," ")))
df_trim$individualID=paste0(sapply(strsplit(df_trim$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",df_trim$sample.name)))
df_trim=df_trim %>% merge(meta[,c("individualID","locationID")],by="individualID")

#Sample names and read counts
df_stat=data.frame(sample.name=read.table(paste0(path,'sample_name.txt'))[1],mapped=read.table(paste0(path,'allCounts.txt'))[1],unmapped=read.table(paste0(path,'unmappedCounts.txt'))[1])
colnames(df_stat)=c('sample.name','mapped','unmapped')
df_stat$individualID=paste0(sapply(strsplit(df_stat$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",df_stat$sample.name)))
df_stat=df_stat %>% merge(meta[,c("individualID","locationID")],by="individualID",all.x=T)
df_stat=df_stat %>% mutate(Ref=sub(".*_", "",sample.name))

#Duplicate stats
dup_stat=list.files(path,"*dup_metrics.txt",full.names = T) %>% lapply(function(x)read.table(x,skip=6,nrows=1,header=T) %>% mutate(sample.name=sub("_marked.*", "",basename(x)))) %>% bind_rows
dup_stat$individualID=paste0(sapply(strsplit(dup_stat$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",dup_stat$sample.name)))
dup_stat=dup_stat %>% mutate(Ref=sub(".*_", "",sample.name))

#Coverage 
cov_stat=list.files(path,"*coverage.txt",full.names = T) %>% lapply(function(x)read.table(x,header=T,comment.char="",col.names=c('rname',"startpos","endpos","numreads","covbases","coverage","meandepth","meanbaseq","meanmapq")) %>% mutate(sample.name=sub("-coverage.*", "",basename(x)))) %>% bind_rows
cov_stat$individualID=paste0(sapply(strsplit(cov_stat$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",cov_stat$sample.name)))
cov_stat=cov_stat %>% mutate(Ref=sub(".*_", "",sample.name))
cov_stat=cov_stat %>% merge(meta[,c("individualID","locationID")],by="individualID",all.x=T)

cov_sum=cov_stat %>% group_by(Ref,individualID) %>% dplyr::summarize(meancov=mean(coverage),sdcov=sd(coverage),meandep=mean(meandepth),sddep=sd(meandepth),meanbq=mean(meanbaseq),sdbq=sd(meanbaseq),meanmq=mean(meanmapq),sdmq=sd(meanmapq))

#flagstats
flag_stat=list.files(path,"*flagstats.txt",full.names = T) %>% lapply(function(x)read.table(x,header=F,sep="\t",col.names=c('value',NA,'metric'))[,c(1,3)]  %>% mutate(sample.name=sub("-flag.*", "",basename(x)))) %>% bind_rows

#Mapped sequence lenght  
index_stat=list.files(path,"*index_stats.txt",full.names = T) %>% lapply(function(x)read.table(x,col.names=c('Ref_name','Seq_length','Mapped_segment','Unmapped_segment')) %>% mutate(sample.name=sub("-index.*", "",basename(x)))) %>% bind_rows
```

#Make global statistics plot 
```{r}
#Percentage of paired reads after trimming with trimmomatic
trim_plot=df_trim %>% subset(Var=="Both Surviving Read Percent")  %>% ggplot(aes(x=individualID,y=100-Value,fill=locationID)) + geom_col(colour='black',position='dodge2') + theme_classic() + labs(x='Sample',y='% Paired reads dropped after trimming') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Percentage of mapped reads to each reference genome 
map_plot=df_stat %>% ggplot(aes(x=individualID,y=100*mapped/(mapped+unmapped),fill=Ref)) + geom_col(colour='black',position='dodge2') + theme_classic() + labs(x='Sample',y='% Mapped reads',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1)) + coord_cartesian(ylim=c(85,100))

#Percentage of duplication per sample 
dup_plot=dup_stat %>% ggplot(aes(x=individualID,y=PERCENT_DUPLICATION,fill=Ref)) + geom_col(colour='black',position='dodge2') + theme_classic() + labs(x='Sample',y='% Duplicated reads',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Coverage
cov_plot=cov_sum %>% ggplot(aes(x=individualID,y=meancov,fill=Ref))   + geom_col(colour='black',position='dodge2') +
  geom_errorbar(aes(ymin=meancov-sdcov, ymax=meancov+sdcov), width=.2,
                 position=position_dodge(.9))+ theme_classic() + labs(x='Sample',y='% Coverage all reference ',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Depth
depth_plot=cov_sum %>% ggplot(aes(x=individualID,y=meandep,fill=Ref))   + geom_col(colour='black',position='dodge2')+ theme_classic() + labs(x='Sample',y='Average depth across refernce chr/contigs',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Mapping quality
mapq_plot=cov_sum %>% ggplot(aes(x=individualID,y=meanmq,fill=Ref))   + geom_col(colour='black',position='dodge2')+ theme_classic() + labs(x='Sample',y='Mapping quality ',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

baseq_plot=cov_sum %>% subset(!duplicated(individualID)) %>% ggplot(aes(x=meanbq))   + geom_histogram(colour='black',binwidth = 0.2,fill="white")+ theme_classic() + labs(x='Base quality',y=' N samples') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1)) + xlim(10,40)

x11()
plot(cov_plot)
```

#Check into details coverage and depth across reference 
```{r}
#MEan coverage for eachr eference across samples 
amil_cov=cov_stat %>% subset(Ref=="Amillepora") %>% group_by(rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meancov,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean coverage across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE)

aspat_cov=cov_stat %>% subset(Ref=="Aspathulata") %>% group_by(rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meancov,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean coverage across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE)

#Link between size of reference genome and coverage
l_plot=cov_stat %>% group_by(Ref,rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% merge(index_stat %>% distinct(Ref_name,.keep_all = T),by.x="rname",by.y="Ref_name") %>% ggplot(aes(x=Seq_length,y=meancov,color=Ref))   + geom_point(size=2)+ theme_classic() + labs(x='log10(Reference length)',y='Mean coverage across samples',color="Reference") + scale_x_continuous(trans='log10')

```


#Save plots 
```{r}
path_save="/home/hugo/PhD/Genomics/Figures/Diagnostic_plot_fasq_to_variant/"

ggsave(trim_plot,filename=paste0(path_save,"Trimming",".png"),width=14,height=10,dpi=320)
ggsave(map_plot,filename=paste0(path_save,"Map_perc",".png"),width=14,height=10,dpi=320)
ggsave(dup_plot,filename=paste0(path_save,"Perc_dup",".png"),width=14,height=10,dpi=320)
ggsave(cov_plot,filename=paste0(path_save,"Coverage",".png"),width=14,height=10,dpi=320)
ggsave(depth_plot,filename=paste0(path_save,"Depth",".png"),width=14,height=10,dpi=320)
ggsave(mapq_plot,filename=paste0(path_save,"Map_q",".png"),width=14,height=10,dpi=320)
ggsave(baseq_plot,filename=paste0(path_save,"Base_q",".png"),width=14,height=10,dpi=320)
ggsave(amil_cov,filename=paste0(path_save,"Amillepora_coverage",".png"),width=14,height=10,dpi=320)
ggsave(aspat_cov,filename=paste0(path_save,"Aspathulata_coverage",".png"),width=14,height=10,dpi=320)
ggsave(l_plot,filename=paste0(path_save,"Ref_coverage~length ",".png"),width=14,height=10,dpi=320)
```

