---
title: "BAM_file_statistics"
author: "Hugo DENIS"
date: "2024-01-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Packages
```{r}
library(openxlsx)
library(dplyr)
library(ggplot2)
library(stringr)
library(tidyr)
```

#### This scripts uses mapped BAM files to produce basic statistics on A.spathulata WGS data ####   
# This includes coverage, base quality, sequencing depth, mapping quality ...

#Load samples metadata (A.spathulata 2022 only)
```{r}
meta=read.csv('/home/hdenis/Data/GBR_data/Coral colony metadata from the ECT1_RRAP_Aspathulata_2022.csv')
```

#Load BAM stats files  
```{r}
path='/nvme/disk0/lecellier_data/WGS_GBR_data/BAM_statistics/'

#Trimmed statistics -> problem with disk 
df_trim=list.files("/tgt_bck2/data_lecellier/Trimmed_files/","*sum.txt",full.names = T) %>% lapply(function(x)read.table(x,sep=":",col.names=c("Var","Value"),colClasses = c("character","numeric")) %>% mutate(sample.name=sub("_sum.*", "",basename(x)))) %>% bind_rows %>% mutate(Value=as.numeric(str_remove(str_replace(Value,",",".")," ")))
# df_trim$individualID=paste0(sapply(strsplit(df_trim$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",df_trim$sample.name)))
# df_trim=df_trim %>% merge(meta[,c("individualID","locationID")],by="individualID")

#Sample names and read counts
df_stat=data.frame(sample.name=read.table(paste0(path,'sample_name.txt'))[1],all=read.table(paste0(path,'allCounts.txt'))[1],unmapped=read.table(paste0(path,'unmappedCounts.txt'))[1])
colnames(df_stat)=c('sample.name','all','unmapped')
# df_stat$individualID=paste0(sapply(strsplit(df_stat$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",df_stat$sample.name)))
# df_stat=df_stat %>% merge(meta[,c("individualID","locationID")],by="individualID",all.x=T)
# df_stat=df_stat %>% mutate(Ref=sub(".*_", "",sample.name))

#Duplicate stats
dup_stat=list.files(path,"*dup_metrics.txt",full.names = T) %>% lapply(function(x)read.table(x,skip=6,nrows=1,header=T) %>% mutate(sample.name=sub("_marked.*", "",basename(x)))) %>% bind_rows
# dup_stat$individualID=paste0(sapply(strsplit(dup_stat$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",dup_stat$sample.name)))
# dup_stat=dup_stat %>% mutate(Ref=sub(".*_", "",sample.name))

#Coverage (genome-wide and chromosomes only )
cov_stat=list.files(path,"*coverage.txt",full.names = T) %>% lapply(function(x)read.table(x,header=T,comment.char="",col.names=c('rname',"startpos","endpos","numreads","covbases","coverage","meandepth","meanbaseq","meanmapq"),colClasses = c("character",rep("numeric",8))) %>% mutate(sample.name=sub("-coverage.*", "",basename(x)))) %>% bind_rows
# cov_stat$individualID=paste0(sapply(strsplit(cov_stat$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",cov_stat$sample.name)))
cov_stat=cov_stat %>% mutate(Ref=sub(".*_", "",sample.name))
# cov_stat=cov_stat %>% merge(meta[,c("individualID","locationID")],by="individualID",all.x=T)

cov_sum_all=cov_stat %>% group_by(sample.name) %>% dplyr::summarize(meancov=mean(coverage),sdcov=sd(coverage),meandep=mean(meandepth),sddep=sd(meandepth),meanbq=mean(meanbaseq),sdbq=sd(meanbaseq),meanmq=mean(meanmapq),sdmq=sd(meanmapq),nreads=sum(numreads))
colnames(cov_sum_all)[2:10]=paste0(colnames(cov_sum_all)[2:10],"_all")

cov_sum_chr=cov_stat %>% subset(rname %in% paste0("scaffold_",seq(1,14))) %>% group_by(sample.name) %>% dplyr::summarize(meancov=mean(coverage),sdcov=sd(coverage),meandep=mean(meandepth),sddep=sd(meandepth),meanbq=mean(meanbaseq),sdbq=sd(meanbaseq),meanmq=mean(meanmapq),sdmq=sd(meanmapq),nreads=sum(numreads))
colnames(cov_sum_chr)[2:10]=paste0(colnames(cov_sum_chr)[2:10],"_chr")

cov_sum=cov_sum_all %>% merge(cov_sum_chr,by="sample.name")

#flagstats
flag_stat=list.files(path,"*flagstats.txt",full.names = T) %>% lapply(function(x)read.table(x,header=F,sep="\t",col.names=c('value',NA,'metric'),colClasses = c("character",NA,"character"))[,c(1,3)]  %>% mutate(sample.name=sub("-flag.*", "",basename(x)))) %>% bind_rows

#Mapped sequence lenght  
index_stat=list.files(path,"*index_stats.txt",full.names = T) %>% lapply(function(x)read.table(x,col.names=c('Ref_name','Seq_length','Mapped_segment','Unmapped_segment'),colClasses =c("character","numeric","numeric","numeric")) %>% mutate(sample.name=sub("-index.*", "",basename(x)))) %>% bind_rows
```

#Make global statistics plot 
```{r}
df_trim =df_trim %>% subset(Var=="Both Surviving Read Percent")
#Percentage of paired reads after trimming with trimmomatic
trim_plot=df_trim %>% mutate(Species=if_else(grepl("2022-Aspat",sample.name),"Aspat",if_else(grepl("2022-Amil",sample.name),"Amil",if_else(grepl("2022-Ahya",sample.name),"Ahya","Unknown")))) %>% ggplot(aes(x=sample.name,y=100-Value,colour=Species,fill=Species)) + geom_col(position='dodge2') + theme_classic() + labs(x='Sample',y='% Paired reads dropped after trimming') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1)) + theme(axis.text.x = element_blank(),axis.text.y = element_text(size=20),axis.title.y = element_text(size=23),axis.title.y.right=element_text(margin=margin(t=0,r=0,b=0,l=10),size=23),axis.title.x = element_text(size=23),legend.text=element_text(size=20),legend.title=element_text(size=20),panel.background =element_rect(fill="white"),plot.margin=margin(t=8,r=4,b=10,l=4,unit="pt"),title=element_text(size=21),panel.border=element_rect(colour="black",fill=NA,size=0.5),legend.key.width = unit(5,"mm"),legend.key.height=unit(15,"mm")) +scale_y_continuous(expand=c(0,0),limits = c(0,15)) + annotate(geom="text",x=df_trim$sample.name[df_trim$Value==min(df_trim$Value)],y=14.7,label=df_trim$sample.name[df_trim$Value==min(df_trim$Value)],size=6) + geom_hline(yintercept =100-median(df_trim$Value),size=2,linetype="dashed",colour="#C60012")

#Percentage of mapped reads to each reference genome 
df_stat=df_stat %>% mutate(perc=100*(all-unmapped)/all)
map_plot=df_stat %>% mutate(Species=if_else(grepl("2022-Aspat",sample.name),"Aspat",if_else(grepl("2022-Amil",sample.name),"Amil",if_else(grepl("2022-Ahya",sample.name),"Ahya","Unknown")))) %>% ggplot(aes(x=sample.name,y=perc,fill=Species,colour=Species)) + geom_col(position='dodge2') + theme_classic() + labs(x='Sample',y='% Mapped reads') + theme(axis.text.x = element_blank(),axis.text.y = element_text(size=20),axis.title.y = element_text(size=23),axis.title.x = element_text(size=23),legend.text=element_text(size=20),legend.title=element_text(size=20),panel.background =element_rect(fill="white"),plot.margin=margin(t=8,r=4,b=10,l=4,unit="pt"),title=element_text(size=21),panel.border=element_rect(colour="black",fill=NA,size=0.5),legend.key.width = unit(5,"mm"),legend.key.height=unit(15,"mm"))+scale_y_continuous(expand=c(0,0)) +coord_cartesian(ylim=c(90,100)) + geom_hline(yintercept =median(df_stat$perc),size=2,linetype="dashed",colour="#C60012")

write.csv(df_stat %>% subset(perc<90),paste0(path,"low_mapping_files.csv"),row.names=F)


#Percentage of duplication per sample 
dup_plot=dup_stat %>% ggplot(aes(x=individualID,y=PERCENT_DUPLICATION,fill=Ref)) + geom_col(colour='black',position='dodge2') + theme_classic() + labs(x='Sample',y='% Duplicated reads',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Coverage, depth, base quality and mapping quality 
stat1=cov_sum %>% subset(grepl("2022-Aspat",sample.name)) %>% dplyr::select(sample.name,meancov_all,meancov_chr,meandep_all,meandep_chr,meanbq_all,meanbq_chr,meanmq_all,meanmq_chr) %>% gather(meancov_all:meanmq_chr,key="Metric",value="Value") %>% group_by(Metric) %>% dplyr::summarise(median=median(Value),sd=sd(Value),min=min(Value),max=max(Value))

stat2=df_stat %>% subset(all>10000000) %>% subset(grepl("2022-Aspat",sample.name)) %>% mutate(mapped=all-unmapped) %>% gather(c(all,mapped),key="Metric",value="Value") %>% group_by(Metric) %>% dplyr::summarise(median=median(Value),sd=sd(Value),min=min(Value),max=max(Value))

stat2$Metric=c("Nreads","Nmapreads")

varname=c("meancov_all" = "Coverage - whole reference","meancov_chr" = "Coverage - chromosomes","meandep_all" = "Depth - whole reference","meandep_chr" = "Depth - chromosomes","meanbq_all" = "Base quality - whole reference","meanbq_chr" = "Base quality - chromosomes","meanmq_all" = "Mapping quality - whole reference","meanmq_chr" = "Mapping quality - chromosomes")

stat1=rbind(stat1,stat2)

write.csv(stat1,paste0(path,"Statistic_summary.csv"),row.names=F)

low_number_sequence=df_stat %>% subset(all-unmapped<10000000)
write.csv(low_number_sequence,paste0(path,"Low_number_seq.csv"),row.names=F)
  

stat_plot=cov_sum %>% subset(grepl("2022-Aspat",sample.name)) %>% dplyr::select(sample.name,meancov_all,meandep_all,meanbq_all,meanmq_all,meancov_chr,meandep_chr,meanbq_chr,meanmq_chr) %>% gather(meancov_all:meanmq_chr,key="Metric",value="Value")%>% ggplot(aes(x=Value)) + geom_histogram(colour="black",fill="white") + theme_classic()+ theme(axis.text.x = element_text(size=15),axis.text.y = element_text(size=15),axis.title.y = element_text(size=23),axis.title.x = element_text(size=23),legend.text=element_text(size=20),legend.title=element_text(size=20),panel.background =element_rect(fill="white"),plot.margin=margin(t=8,r=4,b=10,l=4,unit="pt"),title=element_text(size=21),panel.border=element_rect(colour="black",fill=NA,size=0.5),legend.key.width = unit(5,"mm"),legend.key.height=unit(15,"mm")) + geom_vline(data=median_stat,aes(xintercept =median),size=1,linetype="dashed",colour="#C60012") + geom_text(data=median_stat,aes(x =median,y=300,label=as.character(round(median,1)),hjust=-0.1),size=5,colour="#C60012")+scale_y_continuous(expand=c(0,0))+ facet_wrap(~Metric,scales="free_x",nrow=2,labeller=as_labeller(varname),dir="v",strip.position = "bottom") + theme(strip.background = element_blank(),strip.placement = "outside",strip.text = element_text(size=18)) + labs(x=NULL,y="# samples") + scale_x_continuous(labels=function(x) sprintf("%.1f",x))



cov_plot=cov_sum %>% ggplot(aes(x=individualID,y=meancov,fill=Ref))   + geom_col(colour='black',position='dodge2') +
  geom_errorbar(aes(ymin=meancov-sdcov, ymax=meancov+sdcov), width=.2,
                 position=position_dodge(.9))+ theme_classic() + labs(x='Sample',y='% Coverage all reference ',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Depth
depth_plot=cov_sum %>% ggplot(aes(x=individualID,y=meandep,fill=Ref))   + geom_col(colour='black',position='dodge2')+ theme_classic() + labs(x='Sample',y='Average depth across refernce chr/contigs',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

#Mapping quality
mapq_plot=cov_sum %>% ggplot(aes(x=individualID,y=meanmq,fill=Ref))   + geom_col(colour='black',position='dodge2')+ theme_classic() + labs(x='Sample',y='Mapping quality ',fill='Ref genome') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1))

baseq_plot=cov_sum %>% subset(!duplicated(individualID)) %>% ggplot(aes(x=meanbq))   + geom_histogram(colour='black',binwidth = 0.2,fill="white")+ theme_classic() + labs(x='Base quality',y=' N samples') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1)) + xlim(10,40)

```

#Check into details coverage and depth across reference 
```{r}
#MEan coverage for eachr eference across samples 
amil_cov=cov_stat %>% subset(Ref=="Amillepora") %>% group_by(rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meancov,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean coverage across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE)

aspat_cov=cov_stat %>% subset(Ref=="Aspathulata") %>% group_by(rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meancov,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean coverage across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE)

amil_depth=cov_stat %>% subset(Ref=="Amillepora") %>% group_by(rname) %>% dplyr::summarise(meandepth=mean(meandepth)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meandepth,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean depth across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE) + scale_y_continuous(limits=c(0,20))

amil_histo=cov_stat %>% subset(Ref=="Amillepora") %>% group_by(rname) %>% dplyr::summarise(meandepth=mean(meandepth)) %>% ggplot(aes(x=meandepth))   + geom_histogram(binwidth=10)+ theme_classic() + labs(x='Mean depth across samples',y='N contigs/chr') + guides(color=FALSE)

aspat_depth=cov_stat %>% subset(Ref=="Aspathulata") %>% group_by(rname) %>% dplyr::summarise(meandepth=mean(meandepth)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meandepth,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean depth across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE) + scale_y_continuous(limits=c(0,20))

aspat_histo=cov_stat %>% subset(Ref=="Aspathulata") %>% group_by(rname) %>% dplyr::summarise(meandepth=mean(meandepth)) %>% ggplot(aes(x=meandepth))   + geom_histogram(binwidth=1)+ theme_classic() + labs(x='Mean depth across samples',y='N contigs/chr') + guides(color=FALSE)

aspat_cov=cov_stat %>% subset(Ref=="Aspathulata") %>% group_by(rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% mutate(Ref_type=if_else(str_detect(rname,"CM"),"Chromosome","Contig")) %>% ggplot(aes(x=rname,y=meancov,fill=Ref_type,colour=Ref_type))   + geom_col()+ theme_classic() + labs(x='Reference contig / chr',y='Mean coverage across samples',fill="Reference type") + theme(axis.text.x = element_blank(),axis.ticks.x = element_blank()) + guides(color=FALSE)

#Link between size of reference genome and coverage
l_plot=cov_stat %>% group_by(Ref,rname) %>% dplyr::summarise(meancov=mean(coverage)) %>% merge(index_stat %>% distinct(Ref_name,.keep_all = T),by.x="rname",by.y="Ref_name") %>% ggplot(aes(x=Seq_length,y=meancov,color=Ref))   + geom_point(size=2)+ theme_classic() + labs(x='log10(Reference length)',y='Mean coverage across samples',color="Reference") + scale_x_continuous(trans='log10')

```

 #Check percentage of reads uniquely mapped to chromosome 
```{r}
path='/nvme/disk0/lecellier_data/WGS_GBR_data/Uniquely_mapped_reads_files/'

#Coverage 
cov_stat_uniq=list.files(path,"*coverage.txt",full.names = T) %>% lapply(function(x)read.table(x,header=T,comment.char="",col.names=c('rname',"startpos","endpos","numreads","covbases","coverage","meandepth","meanbaseq","meanmapq")) %>% mutate(sample.name=sub("-coverage.*", "",basename(x)))) %>% bind_rows
cov_stat_uniq$individualID=paste0(sapply(strsplit(cov_stat_uniq$sample.name,"-"),"[[",4),"_",sub("_.*", "",sub(".*-","",cov_stat_uniq$sample.name)))
cov_stat_uniq=cov_stat_uniq %>% mutate(Ref=sub(".*_", "",sample.name))
cov_stat_uniq=cov_stat_uniq %>% merge(meta[,c("individualID","locationID")],by="individualID",all.x=T)


#Some issues check again 
map_chr=cov_stat_uniq %>% group_by(individualID) %>% dplyr::summarize(numreads_uniq=sum(numreads),numreads_chr_uniq=sum(numreads[str_detect(rname,'CM')])) %>% merge(df_stat %>% subset(Ref=="Amillepora") %>% dplyr::select(individualID,all,unmapped),by="individualID") %>% merge(cov_stat %>% subset(Ref=="Amillepora" )%>% group_by(individualID) %>% dplyr::summarize(numreads_chr=sum(numreads[ str_detect(rname,'CM')]),numreads=sum(numreads)) %>% dplyr::select(individualID,numreads,numreads_chr),by="individualID") %>% mutate(Chr_uniq=100*numreads_chr_uniq/all,All=100*numreads/all,Chr=100*numreads_chr/all,All_uniq=100*numreads_uniq/all) %>% gather(c(All,All_uniq,Chr,Chr_uniq),key="Type",value="Percreads") %>%  ggplot(aes(x=individualID,y=Percreads,fill=Type)) + geom_col(colour='black',position='dodge2') + theme_classic() + labs(x='Sample',y='% Mapped reads',fill='Reference') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1)) + coord_cartesian(ylim=c(60,100))

map_chr_loc=cov_stat_uniq %>% group_by(locationID,individualID) %>% dplyr::summarize(numreads_uniq=sum(numreads),numreads_chr_uniq=sum(numreads[str_detect(rname,'CM')])) %>% merge(df_stat %>% subset(Ref=="Amillepora") %>% dplyr::select(individualID,all,unmapped),by="individualID") %>% merge(cov_stat %>% subset(Ref=="Amillepora" )%>% group_by(individualID) %>% dplyr::summarize(numreads_chr=sum(numreads[ str_detect(rname,'CM')]),numreads=sum(numreads)) %>% dplyr::select(individualID,numreads,numreads_chr),by="individualID") %>% group_by(locationID) %>%  dplyr::summarize(Chr_uniq=mean(100*numreads_chr_uniq/all),All=mean(100*numreads/all),Chr=mean(100*numreads_chr/all),All_uniq=mean(100*numreads_uniq/all)) %>% gather(c(All,All_uniq,Chr,Chr_uniq),key="Type",value="Percreads") %>%  ggplot(aes(x=locationID,y=Percreads,fill=Type)) + geom_col(colour='black',position='dodge2') + theme_classic() + labs(x='Samples grouped by location',y='% Mapped reads',fill='Reference') + theme(axis.text.x = element_text(angle=45,vjust = 1, hjust = 1)) + coord_cartesian(ylim=c(60,100))

```


#Save plots 
```{r}
path_save="/home/hdenis/Coral-Genomics/Figures/Diagnostic_plot_fasq_to_variant/"


ggsave(trim_plot,filename=paste0(path_save,"Trimming",".png"),width=16,height=10,dpi=320)
ggsave(map_plot,filename=paste0(path_save,"Map_perc",".png"),width=14,height=10,dpi=320)
ggsave(stat_plot,filename=paste0(path_save,"Stat_sum",".png"),width=18,height=10,dpi=320)


ggsave(dup_plot,filename=paste0(path_save,"Perc_dup",".png"),width=14,height=10,dpi=320)
ggsave(cov_plot,filename=paste0(path_save,"Coverage",".png"),width=14,height=10,dpi=320)
ggsave(depth_plot,filename=paste0(path_save,"Depth",".png"),width=14,height=10,dpi=320)
ggsave(mapq_plot,filename=paste0(path_save,"Map_q",".png"),width=14,height=10,dpi=320)
ggsave(baseq_plot,filename=paste0(path_save,"Base_q",".png"),width=14,height=10,dpi=320)
ggsave(amil_cov,filename=paste0(path_save,"Amillepora_coverage",".png"),width=14,height=10,dpi=320)
ggsave(aspat_cov,filename=paste0(path_save,"Aspathulata_coverage",".png"),width=14,height=10,dpi=320)
ggsave(l_plot,filename=paste0(path_save,"Ref_coverage~length ",".png"),width=14,height=10,dpi=320)

ggsave(amil_depth,filename=paste0(path_save,"Amillepora_depth",".png"),width=14,height=10,dpi=320)
ggsave(aspat_depth,filename=paste0(path_save,"Aspathulata_depth",".png"),width=14,height=10,dpi=320)

ggsave(amil_histo,filename=paste0(path_save,"Amillepora_histo",".png"),width=14,height=10,dpi=320)
ggsave(aspat_histo,filename=paste0(path_save,"Aspathulata_histo",".png"),width=14,height=10,dpi=320)

ggsave(map_chr_loc,filename=paste0(path_save,"Amillepora_mapping_ref_type",".png"),width=14,height=10,dpi=320)
```

