---
title: "Pop_struct_connectivity"
author: "Hugo DENIS"
date: "2024-07-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Packages
```{r}
library(openxlsx)
library(dplyr)
library(ggplot2)
library(stringr)
library(tidyr)
library(cowplot)
library(data.table)
library(geosphere)
```

#### This scripts uses mapped BAM files to produce basic statistics on A.spathulata WGS data ####   
# This includes coverage, base quality, sequencing depth, mapping quality ...

#Load samples metadata (A.spathulata 2022 only)
```{r}
meta=read.csv('/home/hdenis/Data/GBR_data/Coral colony metadata from the ECT1_RRAP_Aspathulata_2022.csv')

site_meta=meta %>% group_by(locationID) %>% dplyr::summarize(Lat=mean(as.numeric(decimalLatitude),na.rm=T),Lon=mean(as.numeric(decimalLongitude),na.rm=T))
```

#Load additional data 
```{r}
fst=list.files('/nvme/disk0/lecellier_data/WGS_GBR_data/Analyses_outputs/Genome_scans/',pattern="pairwise") %>% lapply(function(x) read.table(x)) %>% bind_rows()
```

#Compute pairwise geographic distance between sampled reefs (in km)
```{r}
havdist=data.frame()
for(i in 1:nrow(site_meta)){
  for(j in 1:nrow(site_meta)){
    if(i!=j){
      d=distHaversine(site_meta[i,c(3,2)],site_meta[j,c(3,2)])
      havdist=rbind(havdist,data.frame(Site.name1=site_meta$locationID[i],Site.name2=site_meta$locationID[j],Haversine_dist=d/1000))
    }
}
}
havdist$ID=row.names(havdist)
head(havdist)

havdist %>% ggplot(aes(x=ID,y=Haversine_dist)) + geom_point(size=3)
```


