---
title: "Figure_S6.Rmd"
author: "Hugo DENIS"
date: "2025-06-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

#Packages
```{r}
library(openxlsx)
library(dplyr)
library(ggplot2)
library(stringr)
library(tidyr)
library(cowplot)
library(data.table)
library(adegenet)
library(dartR)
```

#Load data
```{r}
path="C:/Users/Hugo/Documents/Data/Genomics/Aspat_Amil_popgen/"
figpath="C:/Users/Hugo/Documents/Figures/Genomics/Aspat_Amil_popgen/"


#Load global covariance matrix from PCANgsd including all Aspat samples and A.millepora
cov=as.matrix(read.table(paste0(path,"GBR_NC_aspat_amil_samples_05mis_all_chr.cov"))) 

#List files in covariance matrix 
list_files=read.table(paste0(path,'gbr_nc_aspat_noclones_amil.filelist.txt'),col.names='file')
list_files$wgsID_short=sub(".*Aspat-","",sub("_.*","",basename(list_files$file)))

#Get samples metadata
meta_colony=read.table("C:/Users/Hugo/Documents/Data/Colony_metadata_Aspat_GBR_NC.txt",header=T,sep="\t")
meta_site=openxlsx::read.xlsx("C:/Users/Hugo/Documents/Data/Summary_PhD_sites.xlsx",sheet = "Sheet 1")
```

#Plot PCA using hard called variants and a subset of individuals
```{r}
#Get group info for hard called variants
groups=list.files(path,"Group",full.names=T) %>% lapply(function(x)read.table(x,col.names=c("sample")) %>% mutate(Group=sub(".*subset_","",sub(".txt","",basename(x))))) %>% bind_rows()

LD=0.2

#Convert vcf file to genlight object
gbr_nc_genlight=gl.read.vcf(paste0(path,"gbr_nc_aspat_amil_phylo_subset_all_chr_SNP_filtered_2_recoded_chr_nosim_LD_pruned",as.character(LD),".vcf.gz"),verbose=NULL)


Nsites=length(gbr_nc_genlight$loc.names)

cov=glPca(gbr_nc_genlight,parallel=T)

cov.scores=as.data.frame(cov$scores)
cov.scores$wgsID_short=sub(".*Aspat-","",sub("_.*","",rownames(cov.scores)))
cov.scores$sample=rownames(cov.scores)
pca.vectors=cov.scores

#Merge with metadata
pca.vectors=pca.vectors %>% merge(meta_colony %>% dplyr::select(locationID,wgsID_short),by="wgsID_short",all.x=T) %>% mutate(Sample_type=if_else(grepl("ECO03",wgsID_short),"GBR A.millepora",if_else(locationID %in% c("Lady Musgrave","Heron","Fitzroy Reef"),"South GBR A.spathulata",if_else(grepl("ECT01",sample),"Central/North GBR A.spathulata",if_else(grepl("NC",wgsID_short),"NC Taxa1","CB")))))

pca.vectors$Sample_type[pca.vectors$Sample_type=="NC Taxa1" & pca.vectors$PC3>0]="NC Taxa2"

typecolors<-c("#016765","#079CEF","#FFCC33","#CC791E","#CC0000","grey")
names(typecolors)=c("NC Taxa1","CB","South GBR A.spathulata","Central/North GBR A.spathulata","GBR A.millepora","NC Taxa2")


for(i in 1:4){
  perc1=round(100*cov$eig[i]/sum(cov$eig),2)
  perc2=round(100*cov$eig[i+1]/sum(cov$eig),2)
  
  PC1=paste0("PC",as.character(i))
  PC2=paste0("PC",as.character(i+1))
  
  pca = ggplot(data = pca.vectors, aes_string(x=PC1, y=PC2, fill = "Sample_type",color="Sample_type")) + geom_point(size=5.5,alpha=0.8,pch=21) + theme_classic() + theme(axis.text.x = element_text(size=30,colour='black'),axis.text.y = element_text(size=30,colour='black'),axis.title.y = element_text(size=30),axis.title.x = element_text(size=30),legend.text=element_text(size=22),legend.title=element_text(size=25),panel.background =element_rect(fill="white"),plot.margin=margin(t=8,r=4,b=10,l=4,unit="pt"),title=element_text(size=21),panel.border=element_rect(colour="black",fill=NA,size=0.5),legend.key.width = unit(15,"mm"),legend.key.height=unit(20,"mm"),legend.position="right",legend.text.align = 0,legend.background = element_rect(fill="transparent")) + labs(x=paste0(PC1," (",as.character(perc1),"%)"),y=paste0(PC2," (",as.character(perc2),"%)")) + scale_fill_manual(values=typecolors)  + scale_colour_manual(values=typecolors) + guides(fill=guide_legend(override.aes=list(size=14),title="Genomic Cluster"),color="none")
  
  
   ggsave(pca,filename=paste0(figpath,"PCA/Hardcalled_subset_PCA_",PC1,"-",PC2,"_Aspat_Ail_GBR_NC.png"),width=18,height=12,dpi=320)
}
```

#Plot Fst values 
```{r}
library(corrplot)

fst_df=list.files(paste0(path,"Stats"),"*.pairwise.log",full.names=T) %>% lapply(function(x)read.table(x,skip=17,nrow=2)) %>% bind_rows()

fst_df_names=list.files(paste0(path,"Stats"),"*.pairwise.log",full.names=T) %>% lapply(function(x)read.table(x,skip=6,nrow=2)) %>% bind_rows() %>% mutate(POP=sub("_samples.txt","",basename(V2))) %>% mutate(Comparison=rep(seq(1,30),each=2),POP_N=rep(c("POP1","POP2"),30)) %>% dplyr::select(c(POP_N,POP,Comparison)) %>% pivot_wider(names_from="POP_N",values_from = "POP")

fst_df=data.frame(Index="Fst",Type=fst_df$V4,Value=fst_df$V7)

fst_df_mean=cbind(fst_df %>% subset(Type=="mean"),fst_df_names)

#Keep only Group1-4

fst_mat_mean=as.dist(xtabs(fst_df_mean[,3] ~ fst_df_mean[,5] + fst_df_mean[,6]))

fst_mat_mean[upper.tri(fst_mat_mean)]=t(fst_mat_mean[lower.tri(fst_mat_mean,diag=T)])

fst_mat_mean=as.matrix(fst_mat_mean)


#Change group names to match the rest
colnames(fst_mat_mean)=c("CB","GBR A.millepora","Central/North GBR A.spathulata","South GBR A.spathulata","NC Taxa1","NC Taxa2")
rownames(fst_mat_mean)=c("CB","GBR A.millepora","Central/North GBR A.spathulata","South GBR A.spathulata","NC Taxa1","NC Taxa2")

fst_mat_mean=fst_mat_mean[c("GBR A.millepora","NC Taxa2","Central/North GBR A.spathulata","South GBR A.spathulata","CB","NC Taxa1"),c("GBR A.millepora","NC Taxa2","Central/North GBR A.spathulata","South GBR A.spathulata","CB","NC Taxa1")]

typecolors_ordered<-c("#CC0000","grey","#CC791E","#FFCC33","#079CEF","#016765")


png(height=465,width=465,file=paste0(figpath,"Weir_Fst_Group123456.png"),res=300,units="mm"
    )
corrplot::corrplot(as.matrix(fst_mat_mean),col=rev(COL2('RdYlBu',10)),cl.pos='b',tl.cex=2,cl.cex=2,tl.col=typecolors_ordered,col.lim=c(0,0.07),is.corr=F,addCoef.col = 'black',number.cex=2,diag=FALSE,number.digits=4)
dev.off()

```